lib/mongoid/threaded.rb

```diff
@@ -220,6 +224,29 @@ def current_scope=(scope)
       Thread.current[CURRENT_SCOPE_KEY] = scope
     end
 
+    # Set the current Mongoid scope. Safe for multi-model scope chaining.
+    #
+    # @example Set the scope.
+    #   Threaded.current_scope(scope, klass)
+    #
+    # @param [ Criteria ] scope The current scope.
+    # @param [ Klass ] klass The current model class.
+    #
+    # @return [ Criteria ] The scope.
+    #
+    # @since 5.0.1
+    def set_current_scope(scope, klass)
+      if scope.nil?
+        if Thread.current[CURRENT_SCOPE_KEY]
+          Thread.current[CURRENT_SCOPE_KEY].delete(klass)
+          Thread.current[CURRENT_SCOPE_KEY] = nil if Thread.current[CURRENT_SCOPE_KEY].length == 0
+        end
+      else
+        Thread.current[CURRENT_SCOPE_KEY] = {} unless Thread.current[CURRENT_SCOPE_KEY]
```

Comment:
One very ruby-like thing you could do here is

```ruby
Thread.current[CURRENT_SCOPE_KEY] ||= {}
```

This line means set Thread.current[CURRENT_SCOPE_KEY] to {} if it's false (or nil).

It means this: 

```ruby
Thread.current[CURRENT_SCOPE_KEY] || Thread.current[CURRENT_SCOPE_KEY] = {}
```